crates/epaint/src/text/fonts.rs

```diff
@@ -767,20 +774,124 @@ impl GalleyCache {
                 let cached = entry.into_mut();
                 cached.last_used = self.generation;
                 cached.galley.clone()
             }
             std::collections::hash_map::Entry::Vacant(entry) => {
-                let galley = super::layout(fonts, job.into());
-                let galley = Arc::new(galley);
-                entry.insert(CachedGalley {
-                    last_used: self.generation,
-                    galley: galley.clone(),
-                });
-                galley
+                let job = Arc::new(job);
+                if allow_split_paragraphs && should_cache_each_paragraph_individually(&job) {
+                    let galley = self.layout_each_paragraph_individuallly(fonts, job);
+                    let galley = Arc::new(galley);
+                    self.cache.insert(
+                        hash,
+                        CachedGalley {
+                            last_used: self.generation,
+                            galley: galley.clone(),
+                        },
+                    );
+                    galley
+                } else {
+                    let galley = super::layout(fonts, job);
+                    let galley = Arc::new(galley);
+                    entry.insert(CachedGalley {
+                        last_used: self.generation,
+                        galley: galley.clone(),
+                    });
+                    galley
+                }
+            }
+        }
+    }
+
+    /// Split on `\n` and lay out (and cache) each paragraph individually.
+    fn layout_each_paragraph_individuallly(
+        &mut self,
+        fonts: &mut FontsImpl,
+        job: Arc<LayoutJob>,
+    ) -> Galley {
+        profiling::function_scope!();
+
+        let mut start = 0;
+        let mut max_rows_remaining = job.wrap.max_rows;
+        let mut galleys = Vec::new();
+
+        while start < job.text.len() {
+            let is_first_paragraph = start == 0;
+            let end = job.text[start..]
+                .find('\n')
+                .map_or(job.text.len(), |i| start + i + 1);
+
+            let mut paragraph_job = LayoutJob {
+                text: job.text[start..end].to_owned(),
+                wrap: crate::text::TextWrapping {
+                    max_rows: max_rows_remaining,
+                    ..job.wrap
+                },
+                sections: Vec::new(),
+                break_on_newline: job.break_on_newline,
+                halign: job.halign,
+                justify: job.justify,
+                first_row_min_height: if is_first_paragraph {
+                    job.first_row_min_height
+                } else {
+                    0.0
+                },
+                round_output_to_gui: job.round_output_to_gui,
+            };
+
+            // Keep all (previous) sections (so the correct `section_index` is outputted),
+            // but shift their byte ranges:
+            paragraph_job.sections = job
```

Comment:
I changed this code in order to get correct section indices… but this was a mistake. The original idea was _much_ quicker, even though it requires more book-keeping… will need to revert.